content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,756 @@
+
+```go
+func (*File) Diff(ctx context.Context, req infer.DiffRequest[FileArgs, FileState]) (infer.DiffResponse, error) {
+	diff := map[string]p.PropertyDiff{}
+	if req.Inputs.Content != req.State.Content {
+		diff["content"] = p.PropertyDiff{Kind: p.Update}
+	}
+	if req.Inputs.Force != req.State.Force {
+		diff["force"] = p.PropertyDiff{Kind: p.Update}
+	}
+	if req.Inputs.Path != req.State.Path {
+		diff["path"] = p.PropertyDiff{Kind: p.UpdateReplace}
+	}
+	return infer.DiffResponse{
+		DeleteBeforeReplace: true,
+		HasChanges:          len(diff) > 0,
+		DetailedDiff:        diff,
+	}, nil
+}
+```
+
+#### The `Read` function
+
+The `Read` operation fetches the resource. The `ReadRequest` has a `ID` property that can be used, in this case, to determine the path to the file, and the base library functions can be used to read the file from disk, populating the `Content` field.
+
+```go
+func (*File) Read(ctx context.Context, req infer.ReadRequest[FileArgs, FileState]) (infer.ReadResponse[FileArgs, FileState], error) {
+	path := req.ID
+	byteContent, err := os.ReadFile(path)
+	if err != nil {
+		return infer.ReadResponse[FileArgs, FileState]{}, err
+	}
+	content := string(byteContent)
+	return infer.ReadResponse[FileArgs, FileState]{
+		ID: path,
+		Inputs: FileArgs{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
+			Content: content,
+		},
+		State: FileState{
+			Path:    path,
+			Force:   req.Inputs.Force && req.State.Force,
+			Content: content,
+		},
+	}, nil
+}
+```
+
+#### Managing resource output fields
+
+Finally, `WireDependencies` defines the outputs that are made available on the resource, logically connecting the inputs and stored state values.
+
+```go
+func (*File) WireDependencies(f infer.FieldSelector, args *FileArgs, state *FileState) {
+	f.OutputField(&state.Content).DependsOn(f.InputField(&args.Content))
+	f.OutputField(&state.Force).DependsOn(f.InputField(&args.Force))
+	f.OutputField(&state.Path).DependsOn(f.InputField(&args.Path))
+}
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+<!-- TODO: write this section
+
+### Testing with the SDK
+
+ -->
+
+### Multi-language support
+
+In our above example, we created a provider in Go and used it in YAML. This "just works" by default. However, if you would like to suuse your provider from the other Pulumi authoring langauges (e.g. TypeScript, Python, Java, Go, C#) it will be necessary to generate SDKs for each target language.
```

Comment:
```suggestion
In our above example, we created a provider in Go and used it in YAML. This "just works" by default. However, if you would like to use your provider from the other Pulumi authoring langauges (e.g. TypeScript, Python, Java, Go, C#) it will be necessary to generate SDKs for each target language.
```